### ISSUE TITLE
Cmake recipe cannot handle GNU Fortran version 10.x

### ISSUE THREAD

**Issue by chernals** [MethodicalAcceleratorDesign/MAD-X#938]
I was recently thinking about how brittle regexes can be (see https://stackoverflow.com/questions/150764/are-regexes-really-maintainable) and here it comes :)

In `setupGNU.cmake`:

```
if (CMAKE_Fortran_COMPILER_ID MATCHES "GNU")
    # General:
    set(CMAKE_Fortran_FLAGS " -fno-range-check -fno-f2c -cpp ") # remove -g -O2 from main list
    execute_process(COMMAND ${CMAKE_Fortran_COMPILER} --version OUTPUT_VARIABLE CMAKE_Fortran_COMPILER_VERSION)
    string(REGEX MATCH "[0-9].[0-9].[0-9]" CMAKE_Fortran_COMPILER_VERSION ${CMAKE_Fortran_COMPILER_VERSION})
    if(${CMAKE_Fortran_COMPILER_VERSION} VERSION_GREATER 4.3.9)
        add_definitions(-D_GFORTRAN)
    endif()
```

Basically the regex cannot handle versions starting with 10.x.x.

Here is a modified one:

```
    string(REGEX MATCH "1?[0-9].[0-9].[0-9]" CMAKE_Fortran_COMPILER_VERSION ${CMAKE_Fortran_COMPILER_VERSION})
```

It works for version 10, maybe would require some testing for other versions, and we'll have the pleasure to update it when version 20.x are out.

**Comment by chernals** [comment]
The compilation itself fails with GFortran 10 because of invalid BOZ literals. GFortran 10 is closer to the 2008 and 2018 standards and considers it as an error by default.

See lines 202-209 of `util.f90`.

The flag `-fallow-invalid-boz` must be added.

**Comment by tpersson** [comment]
You are right on both points (the later is fixed for the "normal" build system). Thanks for spotting this. Do you think you could do a PR for it and then I test it on a lower version of gnuplot as well?

**Comment by chernals** [comment]
On its way :)

**Comment by tpersson** [comment]
Many thanks!! Fixed in #939

**Comment by Eothred** [comment]
Brilliant, back when I wrote that file GCC was on version 4, and had been for I don't know how many years. I did not dream of reaching double digits at the time :joy:
